emote/callbacks/loss.py

```diff
@@ -44,6 +61,35 @@ def backward(self, *args, **kwargs):
         self.log_scalar(f"loss/{self.name}_loss", loss)
         self.log_scalar(f"loss/{self.name}_gradient_norm", grad_norm)
 
+        if self._log_per_param_weights or self._log_per_param_grads:
+            self.log_per_param_weights_and_grads()
+
+    def log_per_param_weights_and_grads(self):
+        def _friendly_shape_str(shape):
+            return str(list(shape)).replace("[", "").replace("]", "").replace(", ", "_")
```

Comment:
I'd define this outside the function, avoids recreating the function each time. :-)

Reply:
Ah I did not know that was how it worked. Will fix